src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (key in currentState) {
+            const { [key]: _, ...rest } = currentState;
+
+            currentState = rest;
+          }
```

Comment:
You are repeating the action of creating a copy of the currentState in each case of the switch statement. Instead, you can perform this action once after the switch statement to avoid code duplication.